### PR TITLE
SIPP

### PR DESCRIPTION
**Pull request by arbelna**
Implementation of the SIPP algorithm. 
All tests passed (some asserts failed because we don't expand location as a neighboor of itself so the plan sometimes is 1 step shorter).
I also added a new test funcion to both AStar_Solver and AStarSIPP_Solver to check the running time, it looks like the runtime has dropped by about 30 percent for 6*6 map with many constrints

### CODE REVIEW

**Comment by J-morag** [comment]
You've pushed all the commits from your previous PR into this one. Either roll them back in this PR or start a new PR with just your new changes.

**Comment by J-morag** [comment]
> (some asserts failed because we don't expand location as a neighboor of itself so the plan sometimes is 1 step shorter).

This isn't good. It should comply with the exact same contract as the parent class as much as possible.

**Review by J-morag — CHANGES_REQUESTED** [review]
It's very cool that you've managed to implement SIPP while only changing the constraints model, and not the way a state is defined. To the best of my memory, this is not exactly how it was defined in the paper, but seems correct. I've left various comments. Please see them.
Generally, you have many code duplications from the parent class. Avoid them as much as possible. If you only need to make one or two changes to the method, try to factor out those parts of the method into new methods in the base class and then override those. It's okay to copy a method when you make many changes to it.
Finally, please run a longer experiment comparing CBS with A* and CBS with SIPP (offline, not as a new test function) and post the results on runtime and on expanded and generated low level nodes. This would also be a good opportunity to verify that you are always getting the same (optimal) solution cost. The reduction in runtime should be significant.

File: src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+import BasicMAPF.Solvers.ConstraintsAndConflicts.Constraint.I_ConstraintGroupingKey;
+import BasicMAPF.Solvers.Move;
+import BasicMAPF.Solvers.Solution;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.*;
+
+public class SingleAgentAStarSIPP_Solver extends SingleAgentAStar_Solver {
+
+    HashMap<I_Location, ArrayList<int[]>> constraintsByLocation = null;
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java:14]: Done

File: src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+    @Override
+    protected Solution solveAStar() {
+
+        constraintsByLocation = constraintsToTimeIntervals(this.constraints);
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java:19]: Done

File: src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        // this function check if it is possible to wait in place until the destination is
+        // clear of constraints. if it does, it returns the new g value which is equal to the time
+        // the agent waited in place
+        int[] prevRelevantInterval = null;
+        boolean isPossible = false;
+        I_Location prevLocation = state.move.prevLocation;
+        I_Location currLocation = state.move.currLocation;
+        ArrayList<int[]> prevLocationConstraints = constraintsByLocation.get(prevLocation);
+        ArrayList<int[]> currLocationConstraints = constraintsByLocation.get(currLocation);
+        int newProblemStartTime = state.move.timeNow + 1;
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java:148]: Done

File: src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+                if (prevRelevantInterval[1] >= currInterval[0]) {
+                    return -1;
+                }
+            }
+            return prevRelevantInterval[1] + 1;
+        }
+        else return -1;
+    }
+
+    private HashMap<I_Location, ArrayList<int[]>> constraintsToTimeIntervals(ConstraintSet constraints) {
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java:179]: Done

File: src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        // Originally constraints are by location and time, for the SIPP algorithm
+        // we convert the production of the constraints into time intervals by location
+        HashMap<I_Location, ArrayList<int[]>> timeIntervals = new HashMap<>();
+
+        for (I_ConstraintGroupingKey timeLocation : constraints.constraints.keySet()) {
+            if (!timeIntervals.containsKey(timeLocation.getLocation())) {
+                timeIntervals.put(timeLocation.getLocation(), new ArrayList());
+            }
+                ArrayList locationList = timeIntervals.get(timeLocation.getLocation());
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java:188]: Done

File: src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+                locationList.add((timeLocation.getTime()));
+        }
+        for (I_Location location : timeIntervals.keySet()) {
+            ArrayList timestamps = timeIntervals.get(location);
+
+            // Convert the timestamps to intervals using timestampsToIntervals function
+            List<int[]> intervals = timestampsToIntervals(timestamps);
+
+            // Update the output map with the intervals for the current location
+            timeIntervals.put(location, new ArrayList<>(intervals));
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java:198]: It wasn't neccesary, removed the copy

File: src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        }
+
+        return timeIntervals;
+    }
+
+
+    public List<int[]> timestampsToIntervals(List<Integer> timestamps) {
+        // this function gets as input all constraints in a specific location
+        // and return all constraints in timeIntervals instead of timeStamps
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java:207]: Done

File: src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+            int end = timestamp;
+
+            // Extend the interval as long as the next timestamp is one unit away
+            while (i + 1 < timestamps.size() && timestamps.get(i + 1) == end + 1) {
+                end = timestamps.get(i + 1);
+                i++;
+            }
+
+            // Add the interval to the list
+            intervals.add(new int[] {start, end});
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java:229]: I didn't know about record, I changed the data type to be record

File: src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+        List<int[]> intervals = new ArrayList<>();
+        Set<Integer> visited = new HashSet<>();
+
+        for (int i = 0; i < timestamps.size(); i++) {
+            int timestamp = timestamps.get(i);
+
+            // Skip duplicates
+            if (visited.contains(timestamp)) {
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java:217]: I changed the whole function, it begins with sort (which i forgot before) and it's more efficient now

File: src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+            // Start a new interval
+            int start = timestamp;
+            int end = timestamp;
+
+            // Extend the interval as long as the next timestamp is one unit away
+            while (i + 1 < timestamps.size() && timestamps.get(i + 1) == end + 1) {
+                end = timestamps.get(i + 1);
+                i++;
+            }
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java:226]: See comment above

File: src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+            Move lastExistingMove = existingPlan.moveAt(existingPlan.getEndTime());
+            // We assume that we cannot change the existing plan, so if it is rejected by constraints, we can't initialise OPEN.
+            if (constraints.rejects(lastExistingMove)) {
+                return false;
+            }
+
+            openList.add(new AStarState(existingPlan.moveAt(existingPlan.getEndTime()), null, 0, 0));
+        } else { // the existing plan is empty (no existing plan)
+            I_Location sourceLocation = map.getMapLocation(this.sourceCoor);
+
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java:69]: Although it's really not supposed to be happened, i added a check for this in the beginning of the expand function (I guess the location in time 0 is for sure not a NO_STOP location so no need for this check in the init function)

File: src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+            }
+
+            for (I_Location destination : neighborLocationsIncludingCurrent) {
+                Move possibleMove = new Move(agent, problemStartTime + 1, sourceLocation, destination);
+                AStarState rootState = new AStarState(possibleMove, null, this.gAndH.cost(possibleMove), 0);
+                if (constraints.accepts(possibleMove)) { //move not prohibited by existing constraint
+                    openList.add(rootState);
+                } else { // check if it is possible to wait in place until destination will be clear
+//                    rootState = new AStarState(possibleMove, null, problemStartTime + 2, 0);
+                    int isPossible = waitAndMove(rootState);
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java:82]: Done

File: src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+                neighborLocationsIncludingCurrent.add(sourceLocation);
+            }
+
+            for (I_Location destination : neighborLocationsIncludingCurrent) {
+                Move possibleMove = new Move(agent, problemStartTime + 1, sourceLocation, destination);
+                AStarState rootState = new AStarState(possibleMove, null, this.gAndH.cost(possibleMove), 0);
+                if (constraints.accepts(possibleMove)) { //move not prohibited by existing constraint
+                    openList.add(rootState);
+                } else { // check if it is possible to wait in place until destination will be clear
+//                    rootState = new AStarState(possibleMove, null, problemStartTime + 2, 0);
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java:81]: Done

File: src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+                        openList.add(rootState);
+                    }
+                }
+            }
+        }
+        // if none of the root nodes was valid, OPEN will be empty, and thus uninitialised.
+        return !openList.isEmpty();
+    }
+
+    public void expand(@NotNull AStarState state) {
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java:94]: Done

File: src/test/java/BasicMAPF/Solvers/AStar/SingleAgentAStar_SolverTest.java [hunk 15]
```diff
@@ -219,6 +220,35 @@ void circleOptimalityOtherDirectionBecauseOfConstraints(){
...
+        rand.setSeed(10);
+        ConstraintSet constraints = new ConstraintSet();
+        for (int i = 1; i <= 30000; i++) {
+            for (int j = 0; j < 10; j++) {
+
+                I_Location randomLocation = locations.get(rand.nextInt(locations.size()));
+                Constraint constraint = new Constraint(agent, i, null, randomLocation);
+                constraints.add(constraint);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/test/java/BasicMAPF/Solvers/AStar/SingleAgentAStar_SolverTest.java:245]: Done

File: src/test/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_SolverTest.java [hunk 16]
```diff
@@ -0,0 +1,746 @@
...
+        List<I_Location> aStarPlanLocations = new ArrayList<>();
+        for (Move move :
+                planFromAStar) {
+            if (move.timeNow == 1) {
+                aStarPlanLocations.add(move.prevLocation);
+            }
+            aStarPlanLocations.add(move.currLocation);
+        }
+        return aStarPlanLocations;
+    }
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/test/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_SolverTest.java:502]: Done

File: src/test/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_SolverTest.java [hunk 17]
```diff
@@ -0,0 +1,746 @@
...
+        }
+
+        @Override
+        public String toString() {
+//            SortedMap<Edge, Integer> sortedMap = new TreeMap<>(Comparator.comparingInt(e -> randomButStableCosts.get(e)));
+//            sortedMap.putAll(randomButStableCosts);
+            return randomButStableCosts.toString();
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/test/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_SolverTest.java:626]: Done

**Comment by J-morag** [comment]
> It's very cool that you've managed to implement SIPP while only changing the constraints model, and not the way a state is defined. To the best of my memory, this is not exactly how it was defined in the paper, but seems correct. 

I'm having second thoughts about this. I think not adding the safe interval to the sate doesn't lose correctness, but misses some of the speedup, because you're losing out on a lot of duplicate detection. 
Consider if you move from state A at t1 to B at t2. Now at state B, you see that you can get back to state A, at t3. You will generate another state, even if t3 falls within the same safe interval we came from in A. What do you think? Do you agree?
You could still finish this version and leave adding the intervals to the state to future work.

**Comment by arbelna** [comment]
I think you got a point. For now I will continue with the current version
but I will think about how to implement it.

File: src/test/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_SolverTest.java [hunk 18]
```diff
@@ -0,0 +1,746 @@
...
+        Solution expected = new Solution();
+        expected.putPlan(plan);
+
+        assertEquals(5, solved.getPlanFor(agent).size());
+        assertEquals(expected, solved);
+
+    }
+
+    @Test
+    void largeNumberOfConstraints(){
```
[the remark this answers is not in the document]
Comment (arbelna) [reply src/test/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_SolverTest.java:224]: Done

**Comment by J-morag** [comment]
> You've pushed all the commits from your previous PR into this one. Roll them back.

@arbelna Don't forget this.

**Review by arbelna — COMMENTED** [review]
Remove all changes related to outputStreams

**Pull request closed without merging**